File: telegram-bot-sdk/node/src/commands/issue.mjs
/**
 * commands/issue.mjs — Issue tracking Telegram commands
 *
 * Commands:
 *  /issue  [list|add|get|close] [args]  — Issue management
 *  /bug    <title> [description]         — Quick bug report shorthand
 *  /resolve <issue-id>                   — Mark issue as resolved
 */

import { v4 as uuidv4 } from 'uuid';

// ---------------------------------------------------------------------------
// Domain types
// ---------------------------------------------------------------------------

/**
 * @typedef {'open' | 'in-progress' | 'resolved' | 'closed' | 'wontfix'} IssueStatus
 * @typedef {'bug' | 'feature' | 'question' | 'docs' | 'security'} IssueType
 * @typedef {'critical' | 'high' | 'medium' | 'low'} IssuePriority
 */

/**
 * @typedef {Object} Issue
 * @property {string} id
 * @property {string} title
 * @property {string} description
 * @property {IssueStatus} status
 * @property {IssueType} type
 * @property {IssuePriority} priority
 * @property {string} reportedBy  Telegram username or user ID
 * @property {string|null} assignee
 * @property {string} createdAt
 * @property {string|null} resolvedAt
 * @property {string[]} labels
 * @property {Array<{author: string, text: string, timestamp: string}>} comments
 */

// ---------------------------------------------------------------------------
// IssueTracker (in-process store used by IssueCommands)
// ---------------------------------------------------------------------------

class IssueTracker {
  /** @type {Map<string, Issue>} */
  #issues = new Map();

  /** @type {import('../lib/persistence.mjs').PersistenceManager|null} */
  #persistence;

  /** @param {{ persistence?: import('../lib/persistence.mjs').PersistenceManager }} opts */
  constructor({ persistence = null } = {}) {
    this.#persistence = persistence;
    if (persistence) this.#load();
  }

  #save() {
    if (!this.#persistence) return;
    this.#persistence.save('issues', Object.fromEntries(this.#issues));
  }

  #load() {
    const data = this.#persistence.load('issues');
    if (!data) return;
    for (const [k, v] of Object.entries(data)) {
      this.#issues.set(k, v);
    }
  }

  /**
   * @param {{ title: string, description?: string, type?: IssueType,
   *           priority?: IssuePriority, reportedBy: string }} opts
   * @returns {Issue}
   */
  create({ title, description = '', type = 'bug', priority = 'medium', reportedBy }) {
    if (!title?.trim()) throw new Error('Issue title is required');
    const id = uuidv4();
    /** @type {Issue} */
    const issue = {
      id,
      title: title.trim(),
      description: description.trim(),
      status: 'open',
      type,
      priority,
      reportedBy,
      assignee: null,
      createdAt: new Date().toISOString(),
      resolvedAt: null,
      labels: [],
      comments: [],
    };
    this.#issues.set(id, issue);
    this.#save();
    return issue;
  }

  /** @param {string} id @returns {Issue|null} */
  get(id) {
    // Support short IDs (first 8 chars)
    for (const [k, v] of this.#issues) {
      if (k === id || k.startsWith(id)) return v;
    }
    return null;
  }

  /**
   * @param {{ status?: IssueStatus, type?: IssueType, priority?: IssuePriority }} opts
   * @returns {Issue[]}
   */
  list({ status, type, priority } = {}) {
    let issues = Array.from(this.#issues.values());
    if (status) issues = issues.filter((i) => i.status === status);
    if (type) issues = issues.filter((i) => i.type === type);
    if (priority) issues = issues.filter((i) => i.priority === priority);
    return issues.sort(
      (a, b) => new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime()
    );
  }

  /** @param {string} id @returns {Issue} */
  resolve(id) {
    const issue = this.#require(id);
    issue.status = 'resolved';
    issue.resolvedAt = new Date().toISOString();
    this.#save();
    return issue;
  }

  /** @param {string} id @returns {Issue} */
  close(id) {
    const issue = this.#require(id);
    issue.status = 'closed';
    this.#save();
    return issue;
  }

  /** @param {string} id @returns {Issue} */
  reopen(id) {
    const issue = this.#require(id);
    issue.status = 'open';
    issue.resolvedAt = null;
    this.#save();
    return issue;
  }

  /**
   * @param {string} id
   * @param {Partial<Issue>} patch
   * @returns {Issue}
   */
  update(id, patch) {
    const issue = this.#require(id);
    const forbidden = ['id', 'createdAt', 'reportedBy'];
    for (const k of forbidden) delete patch[k];
    Object.assign(issue, patch);
    this.#save();
    return issue;
  }

  /**
   * @param {string} id
   * @param {{ author: string, text: string }} comment
   * @returns {Issue}
   */
  addComment(id, { author, text }) {
    const issue = this.#require(id);
    issue.comments.push({ author, text, timestamp: new Date().toISOString() });
    this.#save();
    return issue;
  }

  #require(id) {
    const issue = this.get(id);
    if (!issue) throw new Error(`Issue not found: ${id}`);
    return issue;
  }
}

// ---------------------------------------------------------------------------
// Formatting helpers
// ---------------------------------------------------------------------------

const STATUS_EMOJI = {
  open: '🔴',
  'in-progress': '🔵',
  resolved: '✅',
  closed: '⬛',
  wontfix: '🚫',
};

const PRIORITY_EMOJI = {
  critical: '🚨',
  high: '🔥',
  medium: '⚠️',
  low: '💡',
};

const TYPE_EMOJI = {
  bug: '🐛',
  feature: '✨',
  question: '❓',
  docs: '📄',
  security: '🔒',
};

function escMd(text) {
  return String(text).replace(/[_*[\]()~`>#+=|{}.!\\-]/g, (c) => `\\${c}`);
}

function fmtIssue(issue, short = false) {
  const se = STATUS_EMOJI[issue.status] ?? '•';
  const pe = PRIORITY_EMOJI[issue.priority] ?? '';
  const te = TYPE_EMOJI[issue.type] ?? '';
  const id = `\`${issue.id.slice(0, 8)}\``;

  if (short) {
    return `${se}${pe}${te} ${escMd(issue.title)} ${id}`;
  }

  const assignee = issue.assignee ? `\nAssignee: @${escMd(issue.assignee)}` : '';
  const reporter = `\nReporter: ${escMd(issue.reportedBy)}`;
  const created = `\nCreated: ${issue.createdAt.slice(0, 10)}`;
  const resolved = issue.resolvedAt
    ? `\nResolved: ${issue.resolvedAt.slice(0, 10)}`
    : '';
  const desc = issue.description
    ? `\n\n${escMd(issue.description.slice(0, 300))}`
    : '';

  return (
    `${se} *${escMd(issue.title)}* ${id}\n` +
    `Type: ${te} ${issue.type} | Priority: ${pe} ${issue.priority}${assignee}${reporter}${created}${resolved}${desc}`
  );
}

// ---------------------------------------------------------------------------
// IssueCommands
// ---------------------------------------------------------------------------

export class IssueCommands {
  #tracker;

  /** @param {{ persistence?: import('../lib/persistence.mjs').PersistenceManager }} opts */
  constructor({ persistence } = {}) {
    this.#tracker = new IssueTracker({ persistence });
  }

  /**
   * Register all issue commands with a Telegraf bot.
   * @param {import('telegraf').Telegraf} bot
   */
  register(bot) {
    bot.command('issue', (ctx) => this.#handleIssue(ctx));
    bot.command('bug', (ctx) => this.#handleBug(ctx));
    bot.command('resolve', (ctx) => this.#handleResolve(ctx));
  }

  // -------------------------------------------------------------------------
  // /issue
  // -------------------------------------------------------------------------

  async #handleIssue(ctx) {
    const rawText = ctx.message?.text ?? '';
    const parts = rawText.replace(/^\/\w+\s*/, '').trim().split(/\s+/);
    const sub = parts[0]?.toLowerCase() ?? '';

    try {
      switch (sub) {
        case 'add':
        case 'new': {
          const title = parts.slice(1).join(' ').trim();
          if (!title) return ctx.reply('Usage: `/issue add <title>`', { parse_mode: 'Markdown' });
          const reporter = ctx.from?.username ?? String(ctx.from?.id ?? 'unknown');
          const issue = this.#tracker.create({ title, reportedBy: reporter });
          return ctx.reply(
            `✅ Issue created\\!\n\n${fmtIssue(issue)}`,
            { parse_mode: 'Markdown' }
          );
        }

        case 'get':
        case 'show': {
          const id = parts[1];
          if (!id) return ctx.reply('Usage: `/issue get <id>`', { parse_mode: 'Markdown' });
          const issue = this.#tracker.get(id);
          if (!issue) return ctx.reply(`Issue not found: ${id}`);
          return ctx.reply(fmtIssue(issue), { parse_mode: 'Markdown' });
        }

        case 'close': {
          const id = parts[1];
          if (!id) return ctx.reply('Usage: `/issue close <id>`', { parse_mode: 'Markdown' });
          const issue = this.#tracker.close(id);
          return ctx.reply(`⬛ Issue *${escMd(issue.title)}* closed.`, { parse_mode: 'Markdown' });
        }

        case 'reopen': {
          const id = parts[1];
          if (!id) return ctx.reply('Usage: `/issue reopen <id>`', { parse_mode: 'Markdown' });
          const issue = this.#tracker.reopen(id);
          return ctx.reply(`🔴 Issue *${escMd(issue.title)}* reopened.`, { parse_mode: 'Markdown' });
        }

        case 'assign': {
          const id = parts[1];
          const assignee = parts[2]?.replace('@', '');
          if (!id || !assignee) return ctx.reply('Usage: `/issue assign <id> <@user>`', { parse_mode: 'Markdown' });
          const issue = this.#tracker.update(id, { assignee });
          return ctx.reply(`👤 Issue assigned to @${escMd(assignee)}.`, { parse_mode: 'Markdown' });
        }

        case 'list':
        case '': {
          const statusFilter = parts[1];
          const validStatuses = ['open', 'in-progress', 'resolved', 'closed', 'wontfix'];
          const status = validStatuses.includes(statusFilter) ? statusFilter : 'open';
          const issues = this.#tracker.list({ status });

          if (issues.length === 0) {
            return ctx.reply(`No ${status} issues found.`);
          }

          const lines = issues
            .slice(0, 15)
            .map((i) => fmtIssue(i, true));
          const more = issues.length > 15 ? `\n_...and ${issues.length - 15} more_` : '';
          return ctx.reply(
            `*Issues (${status})* — ${issues.length} total\n\n${lines.join('\n')}${more}`,
            { parse_mode: 'Markdown' }
          );
        }

        default:
          return ctx.reply(this.#issueHelp(), { parse_mode: 'Markdown' });
      }
    } catch (err) {
      return ctx.reply(`❌ ${err.message}`);
    }
  }

  // -------------------------------------------------------------------------
  // /bug
  // -------------------------------------------------------------------------

  async #handleBug(ctx) {
    const rawText = ctx.message?.text ?? '';
    const content = rawText.replace(/^\/\w+\s*/, '').trim();

    if (!content) {
      return ctx.reply(
        '🐛 *Report a bug*\n\nUsage: `/bug <title> | <description>`\n\nExample:\n`/bug Login fails on mobile | The login button is unresponsive on iOS 17`',
        { parse_mode: 'Markdown' }
      );
    }

    const [title, ...descParts] = content.split('|');
    const description = descParts.join('|').trim();
    const reporter = ctx.from?.username ?? String(ctx.from?.id ?? 'unknown');

    try {
      const issue = this.#tracker.create({
        title: title.trim(),
        description,
        type: 'bug',
        priority: 'high',
        reportedBy: reporter,
      });

      return ctx.reply(
        `🐛 Bug reported\\!\n\n${fmtIssue(issue)}`,
        { parse_mode: 'Markdown' }
      );
    } catch (err) {
      return ctx.reply(`❌ ${err.message}`);
    }
  }

  // -------------------------------------------------------------------------
  // /resolve
  // -------------------------------------------------------------------------

  async #handleResolve(ctx) {
    const rawText = ctx.message?.text ?? '';
    const id = rawText.replace(/^\/\w+\s*/, '').trim();

    if (!id) {
      return ctx.reply('Usage: `/resolve <issue-id>`', { parse_mode: 'Markdown' });
    }

    try {
      const issue = this.#tracker.resolve(id);
      return ctx.reply(
        `✅ Issue *${escMd(issue.title)}* resolved\\!`,
        { parse_mode: 'Markdown' }
      );
    } catch (err) {
      return ctx.reply(`❌ ${err.message}`);
    }
  }

  // -------------------------------------------------------------------------
  // Help
  // -------------------------------------------------------------------------

  #issueHelp() {
    return (
      `*Issue Tracking Commands*\n\n` +
      `\`/issue list [status]\` — List issues (default: open)\n` +
      `\`/issue add <title>\` — Create a new issue\n` +
      `\`/issue get <id>\` — Show issue details\n` +
      `\`/issue close <id>\` — Close an issue\n` +
      `\`/issue reopen <id>\` — Reopen an issue\n` +
      `\`/issue assign <id> @user\` — Assign issue\n` +
      `\`/bug <title> [| description]\` — Quick bug report\n` +
      `\`/resolve <id>\` — Resolve an issue`
    );
  }
}

export default IssueCommands;

File: telegram-bot-sdk/rust/src/commands/issue.rs
//! Issue tracking command handlers.
//!
//! Commands:
//!  - `/issue create <title> [--priority low|medium|high|critical] [--points N]`
//!  - `/issue list [--sprint id] [--status status] [--assignee user]`
//!  - `/issue update <id-prefix> [--status ...] [--assignee ...] [--points N]`
//!  - `/issue close <id-prefix>`
//!  - `/issue show <id-prefix>`

use std::sync::Arc;
use async_trait::async_trait;
use tracing::info;

use crate::commands::Command;
use crate::error::{BotError, Result};
use crate::state::{AppState, Issue, IssueStatus, IssuePriority};
use crate::types::{Context, Response};

// ---------------------------------------------------------------------------
// IssueCommand — /issue
// ---------------------------------------------------------------------------

pub struct IssueCommand {
    state: Arc<AppState>,
}

impl IssueCommand {
    pub fn new(state: Arc<AppState>) -> Self {
        Self { state }
    }
}

#[async_trait]
impl Command for IssueCommand {
    fn name(&self) -> &str { "issue" }
    fn description(&self) -> &str { "Issue tracking" }
    fn usage(&self) -> &str { "[create|list|update|close|show] [args]" }

    async fn execute(&self, ctx: &Context, args: &[&str]) -> Result<Response> {
        let sub = args.first().copied().unwrap_or("list");
        let rest = if args.len() > 1 { &args[1..] } else { &[] as &[&str] };
        match sub {
            "create" | "new" | "add" => self.create(ctx, rest).await,
            "list" | "ls" => self.list(ctx, rest).await,
            "update" | "edit" => self.update(ctx, rest).await,
            "close" | "done" | "complete" => self.close(ctx, rest).await,
            "show" | "info" | "get" => self.show(ctx, rest).await,
            "assign" => self.assign(ctx, rest).await,
            "help" | _ => Ok(Response::text(self.issue_help())),
        }
    }
}

impl IssueCommand {
    // -----------------------------------------------------------------------
    // create
    // -----------------------------------------------------------------------

    async fn create(&self, ctx: &Context, args: &[&str]) -> Result<Response> {
        if args.is_empty() {
            return Ok(Response::text(
                "Usage: /issue create <title> [--priority low|medium|high|critical] [--points N] [--sprint id]\n\
                 Example: /issue create \"Fix login bug\" --priority high --points 3",
            ));
        }

        // First non-flag arg is the title.
        let mut title_parts = Vec::new();
        let mut priority = IssuePriority::Medium;
        let mut story_points: Option<u32> = None;
        let mut sprint_prefix: Option<&str> = None;
        let mut assignee: Option<&str> = None;

        let mut i = 0;
        while i < args.len() {
            match args[i] {
                "--priority" | "-p" => {
                    i += 1;
                    priority = parse_priority(args.get(i).copied().unwrap_or("medium"));
                }
                "--points" | "--sp" => {
                    i += 1;
                    story_points = args.get(i).and_then(|s| s.parse().ok());
                }
                "--sprint" | "-s" => {
                    i += 1;
                    sprint_prefix = args.get(i).copied();
                }
                "--assignee" | "-a" => {
                    i += 1;
                    assignee = args.get(i).copied();
                }
                other => title_parts.push(other),
            }
            i += 1;
        }

        let title = title_parts.join(" ");
        if title.trim().is_empty() {
            return Err(BotError::CommandError("Issue title cannot be empty".to_string()));
        }

        let mut issue = Issue::new(title.trim());
        issue.priority = priority;
        issue.story_points = story_points;
        issue.reporter = ctx.user.as_ref().and_then(|u| u.username.clone());
        issue.assignee = assignee.map(|a| a.trim_start_matches('@').to_string());

        // Resolve sprint
        if let Some(prefix) = sprint_prefix {
            let sprints = self.state.list_sprints().await;
            if let Some(s) = sprints.iter().find(|s| s.id.to_string().starts_with(prefix)) {
                issue.sprint_id = Some(s.id);
            }
        } else if let Some(active) = self.state.get_active_sprint().await {
            issue.sprint_id = Some(active.id);
        }

        let id_short = issue.id.to_string()[..8].to_owned();
        let created = self.state.create_issue(issue).await?;

        // Add the issue to the sprint's issue list
        if let Some(sid) = created.sprint_id {
            let _ = self.state.update_sprint(sid, |s| {
                s.add_issue(created.id);
            }).await;
        }

        info!("Issue created: {} ({})", created.title, created.id);
        Ok(Response::text(format!(
            "Issue created!\n\
             ID: {}\n\
             Title: {}\n\
             Priority: {}\n\
             Points: {}\n\
             Sprint: {}",
            id_short,
            created.title,
            created.priority,
            created.story_points.map(|p| p.to_string()).unwrap_or_else(|| "-".to_string()),
            created.sprint_id.map(|id| id.to_string()[..8].to_string()).unwrap_or_else(|| "none".to_string()),
        )))
    }

    // -----------------------------------------------------------------------
    // list
    // -----------------------------------------------------------------------

    async fn list(&self, _ctx: &Context, args: &[&str]) -> Result<Response> {
        let mut sprint_filter: Option<String> = None;
        let mut status_filter: Option<IssueStatus> = None;
        let mut assignee_filter: Option<String> = None;

        let mut i = 0;
        while i < args.len() {
            match args[i] {
                "--sprint" | "-s" => {
                    i += 1;
                    sprint_filter = args.get(i).map(|s| s.to_string());
                }
                "--status" => {
                    i += 1;
                    status_filter = args.get(i).map(|s| parse_status(s));
                }
                "--assignee" | "-a" => {
                    i += 1;
                    assignee_filter = args.get(i).map(|s| s.trim_start_matches('@').to_string());
                }
                _ => {}
            }
            i += 1;
        }

        let all_issues = self.state.list_issues().await;
        let mut issues: Vec<_> = all_issues.iter().collect();

        // Apply sprint filter
        if let Some(ref prefix) = sprint_filter {
            let sprints = self.state.list_sprints().await;
            if let Some(sprint) = sprints.iter().find(|s| {
                s.id.to_string().starts_with(prefix.as_str()) || s.name == prefix.as_str()
            }) {
                let sid = sprint.id;
                issues.retain(|i| i.sprint_id == Some(sid));
            }
        } else {
            // Default: show active sprint issues if available
            if let Some(active) = self.state.get_active_sprint().await {
                if sprint_filter.is_none() && status_filter.is_none() {
                    issues.retain(|i| i.sprint_id == Some(active.id));
                }
            }
        }

        // Apply status filter
        if let Some(ref status) = status_filter {
            issues.retain(|i| &i.status == status);
        }

        // Apply assignee filter
        if let Some(ref assignee) = assignee_filter {
            issues.retain(|i| {
                i.assignee.as_deref().map(|a| a == assignee.as_str()).unwrap_or(false)
            });
        }

        if issues.is_empty() {
            return Ok(Response::text("No issues found."));
        }

        let mut lines = vec![format!("Issues ({}):", issues.len())];
        for issue in issues {
            let status_icon = status_icon(&issue.status);
            let pts = issue.story_points.map(|p| format!("[{}pt]", p)).unwrap_or_default();
            let assignee = issue.assignee.as_deref().map(|a| format!(" @{}", a)).unwrap_or_default();
            lines.push(format!(
                "{} [{}] {} {}{}",
                status_icon,
                &issue.id.to_string()[..8],
                issue.title,
                pts,
                assignee,
            ));
        }
        Ok(Response::text(lines.join("\n")))
    }

    // -----------------------------------------------------------------------
    // update
    // -----------------------------------------------------------------------

    async fn update(&self, _ctx: &Context, args: &[&str]) -> Result<Response> {
        let id_prefix = args.first().copied().ok_or_else(|| {
            BotError::CommandError("Usage: /issue update <id> [--status ...] [--assignee ...] [--points N]".to_string())
        })?;

        let issue = self
            .state
            .find_issue_by_prefix(id_prefix)
            .await
            .ok_or_else(|| BotError::StateError(format!("Issue '{}' not found", id_prefix)))?;

        let mut new_status: Option<IssueStatus> = None;
        let mut new_assignee: Option<String> = None;
        let mut new_points: Option<u32> = None;
        let mut new_priority: Option<IssuePriority> = None;

        let mut i = 1;
        while i < args.len() {
            match args[i] {
                "--status" | "-s" => {
                    i += 1;
                    new_status = args.get(i).map(|s| parse_status(s));
                }
                "--assignee" | "-a" => {
                    i += 1;
                    new_assignee = args.get(i).map(|s| s.trim_start_matches('@').to_string());
                }
                "--points" | "--sp" => {
                    i += 1;
                    new_points = args.get(i).and_then(|s| s.parse().ok());
                }
                "--priority" | "-p" => {
                    i += 1;
                    new_priority = args.get(i).map(|s| parse_priority(s));
                }
                _ => {}
            }
            i += 1;
        }

        let updated = self.state.update_issue(issue.id, |iss| {
            if let Some(status) = new_status.take() {
                iss.status = status;
            }
            if let Some(assignee) = new_assignee.take() {
                iss.assignee = Some(assignee);
            }
            if let Some(pts) = new_points.take() {
                iss.story_points = Some(pts);
            }
            if let Some(priority) = new_priority.take() {
                iss.priority = priority;
            }
        }).await?;

        Ok(Response::text(format!(
            "Issue updated: [{}] {}\nStatus: {}\nPriority: {}\nAssignee: {}\nPoints: {}",
            &updated.id.to_string()[..8],
            updated.title,
            updated.status,
            updated.priority,
            updated.assignee.as_deref().unwrap_or("unassigned"),
            updated.story_points.map(|p| p.to_string()).unwrap_or_else(|| "-".to_string()),
        )))
    }

    // -----------------------------------------------------------------------
    // close
    // -----------------------------------------------------------------------

    async fn close(&self, _ctx: &Context, args: &[&str]) -> Result<Response> {
        let id_prefix = args.first().copied().ok_or_else(|| {
            BotError::CommandError("Usage: /issue close <id>".to_string())
        })?;

        let issue = self
            .state
            .find_issue_by_prefix(id_prefix)
            .await
            .ok_or_else(|| BotError::StateError(format!("Issue '{}' not found", id_prefix)))?;

        let updated = self.state.update_issue(issue.id, |i| {
            i.status = IssueStatus::Done;
        }).await?;

        info!("Issue closed: {} ({})", updated.title, updated.id);
        Ok(Response::text(format!(
            "Issue closed: [{}] {}",
            &updated.id.to_string()[..8],
            updated.title,
        )))
    }

    // -----------------------------------------------------------------------
    // show
    // -----------------------------------------------------------------------

    async fn show(&self, _ctx: &Context, args: &[&str]) -> Result<Response> {
        let id_prefix = args.first().copied().ok_or_else(|| {
            BotError::CommandError("Usage: /issue show <id>".to_string())
        })?;

        let issue = self
            .state
            .find_issue_by_prefix(id_prefix)
            .await
            .ok_or_else(|| BotError::StateError(format!("Issue '{}' not found", id_prefix)))?;

        Ok(Response::text(format!(
            "Issue: {}\n\
             ID: {}\n\
             Status: {}\n\
             Priority: {}\n\
             Story Points: {}\n\
             Assignee: {}\n\
             Reporter: {}\n\
             Sprint: {}\n\
             Labels: {}\n\
             Created: {}\n\
             Updated: {}\n\
             {}",
            issue.title,
            &issue.id.to_string()[..8],
            issue.status,
            issue.priority,
            issue.story_points.map(|p| p.to_string()).unwrap_or_else(|| "—".to_string()),
            issue.assignee.as_deref().unwrap_or("unassigned"),
            issue.reporter.as_deref().unwrap_or("unknown"),
            issue.sprint_id.map(|id| id.to_string()[..8].to_string()).unwrap_or_else(|| "none".to_string()),
            if issue.labels.is_empty() { "none".to_string() } else { issue.labels.join(", ") },
            issue.created_at.format("%Y-%m-%d %H:%M UTC"),
            issue.updated_at.format("%Y-%m-%d %H:%M UTC"),
            issue.description.as_deref().map(|d| format!("\n{}", d)).unwrap_or_default(),
        )))
    }

    // -----------------------------------------------------------------------
    // assign
    // -----------------------------------------------------------------------

    async fn assign(&self, _ctx: &Context, args: &[&str]) -> Result<Response> {
        let id_prefix = args.first().copied().ok_or_else(|| {
            BotError::CommandError("Usage: /issue assign <id> <@username>".to_string())
        })?;
        let assignee = args.get(1).copied().ok_or_else(|| {
            BotError::CommandError("Usage: /issue assign <id> <@username>".to_string())
        })?;

        let issue = self
            .state
            .find_issue_by_prefix(id_prefix)
            .await
            .ok_or_else(|| BotError::StateError(format!("Issue '{}' not found", id_prefix)))?;

        let clean_assignee = assignee.trim_start_matches('@').to_string();
        let updated = self.state.update_issue(issue.id, |i| {
            i.assignee = Some(clean_assignee.clone());
        }).await?;

        Ok(Response::text(format!(
            "Assigned [{}] {} to @{}",
            &updated.id.to_string()[..8],
            updated.title,
            clean_assignee,
        )))
    }

    fn issue_help(&self) -> String {
        "Issue Commands:\n\
         /issue create <title> [--priority p] [--points N] — Create issue\n\
         /issue list [--sprint id] [--status s] [--assignee a] — List issues\n\
         /issue show <id> — Show issue details\n\
         /issue update <id> [--status s] [--assignee a] [--points N] — Update\n\
         /issue close <id> — Mark as done\n\
         /issue assign <id> <@user> — Assign to user"
            .to_string()
    }
}

// ---------------------------------------------------------------------------
// IssueDispatcher — convenience facade used by bot.rs
// ---------------------------------------------------------------------------

pub struct IssueDispatcher {
    state: Arc<AppState>,
}

impl IssueDispatcher {
    pub fn new(state: Arc<AppState>) -> Self {
        Self { state }
    }

    pub async fn dispatch(&self, ctx: &Context, args: &[&str]) -> Result<Response> {
        IssueCommand::new(self.state.clone()).execute(ctx, args).await
    }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

fn parse_priority(s: &str) -> IssuePriority {
    match s.to_ascii_lowercase().as_str() {
        "low" | "l" => IssuePriority::Low,
        "high" | "h" => IssuePriority::High,
        "critical" | "crit" | "c" => IssuePriority::Critical,
        _ => IssuePriority::Medium,
    }
}

fn parse_status(s: &str) -> IssueStatus {
    match s.to_ascii_lowercase().replace('-', "").replace('_', "").as_str() {
        "todo" => IssueStatus::Todo,
        "inprogress" | "progress" | "wip" => IssueStatus::InProgress,
        "inreview" | "review" => IssueStatus::InReview,
        "done" | "complete" | "completed" => IssueStatus::Done,
        "cancelled" | "canceled" => IssueStatus::Cancelled,
        _ => IssueStatus::Backlog,
    }
}

fn status_icon(status: &IssueStatus) -> &'static str {
    match status {
        IssueStatus::Backlog => "  ",
        IssueStatus::Todo => "[ ]",
        IssueStatus::InProgress => "[~]",
        IssueStatus::InReview => "[R]",
        IssueStatus::Done => "[x]",
        IssueStatus::Cancelled => "[/]",
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::state::BotConfig;
    use crate::types::Message;

    fn make_state() -> Arc<AppState> {
        AppState::new(BotConfig { token: "test".into(), ..Default::default() })
    }

    fn ctx() -> Context {
        Context::new(Message::new(1, 1, "/issue"))
    }

    #[tokio::test]
    async fn create_and_list() {
        let state = make_state();
        let cmd = IssueCommand::new(state.clone());

        cmd.execute(&ctx(), &["create", "Bug in auth", "--priority", "high"])
            .await
            .unwrap();
        cmd.execute(&ctx(), &["create", "Refactor DB", "--points", "5"])
            .await
            .unwrap();

        // List without active sprint returns all when filtering is off
        let resp = cmd.execute(&ctx(), &["list", "--status", "backlog"]).await.unwrap();
        assert!(resp.as_str().contains("Bug in auth") || resp.as_str().contains("Refactor DB"));
    }

    #[tokio::test]
    async fn close_issue() {
        let state = make_state();
        let cmd = IssueCommand::new(state.clone());

        cmd.execute(&ctx(), &["create", "Flaky test"]).await.unwrap();
        let issues = state.list_issues().await;
        let prefix = &issues[0].id.to_string()[..8];

        let resp = cmd.execute(&ctx(), &["close", prefix]).await.unwrap();
        assert!(resp.as_str().contains("closed"));
    }

    #[test]
    fn parse_priority_cases() {
        assert_eq!(parse_priority("critical"), IssuePriority::Critical);
        assert_eq!(parse_priority("LOW"), IssuePriority::Low);
        assert_eq!(parse_priority("anything"), IssuePriority::Medium);
    }
}
